File: qbopt/backend/lower_floats.py
"""Refuse floating operations the backend cannot encode."""

from dataclasses import replace

from qbopt.frontend import raising_floats
from qbopt.model import mir


def _removed(op: mir.Op) -> bool:
    from qbopt.backend.lower import Unlowered

    if op.kind not in (mir.Kind.NOTHING, mir.Kind.FCHECK):
        return False
    if op.floating is not None or op.stack is not None:
        raise Unlowered("removed floating operation retains computation")
    return True


def checked(body: mir.MirBody) -> None:
    from qbopt.backend.lower import Unlowered

    repetitions = dict(body.repetitions)
    if len(repetitions) != len(body.repetitions) or any(
        body.block(at) is None or not 2 <= count <= len(body.block(at).ops) for at, count in body.repetitions
    ):
        raise Unlowered("invalid block repetition provenance")
    for block in body.blocks:
        for op in block.ops:
            if _removed(op) or op.floating is None:
                continue
            # Selection encodes the instruction's name and operands, never
            # `floating`; only the exception policy is not in the bytes.
            encoded = raising_floats.semantics(op)
            if encoded is None or replace(encoded, exceptions=op.floating.exceptions) != op.floating:
                raise Unlowered("floating semantics are not what the instruction encodes")

File: src/backend/lower_floats.rs
//! Port of `qbopt/backend/lower_floats.py`: refuse floating operations the
//! backend cannot encode.

use crate::backend::lower::Unlowered;
use crate::frontend::raising_floats;
use crate::model::mir::{Kind, MirBody, Op};

fn _removed(op: &Op) -> Result<bool, Unlowered> {
    if !matches!(op.kind, Kind::Nothing | Kind::Fcheck) {
        return Ok(false);
    }
    if op.floating.is_some() || op.stack.is_some() {
        return Err(Unlowered("removed floating operation retains computation".into()));
    }
    Ok(true)
}

pub fn checked(body: &MirBody) -> Result<(), Unlowered> {
    let repetitions: crate::support::hash::HashMap<i64, i64> = body.repetitions.iter().copied().collect();
    if repetitions.len() != body.repetitions.len()
        || body.repetitions.iter().any(|&(at, count)| {
            body.block(at).is_none_or(|block| !(2 <= count && count <= block.ops.len() as i64))
        })
    {
        return Err(Unlowered("invalid block repetition provenance".into()));
    }
    for block in &body.blocks {
        for op in &block.ops {
            if _removed(op)? {
                continue;
            }
            let Some(floating) = &op.floating else {
                continue;
            };
            // Selection encodes the instruction's name and operands, never
            // `floating`; only the exception policy is not in the bytes.
            let encoded = raising_floats::semantics(op);
            if encoded.is_none_or(|encoded| {
                crate::model::floating::Semantics { exceptions: floating.exceptions, ..encoded } != *floating
            }) {
                return Err(Unlowered("floating semantics are not what the instruction encodes".into()));
            }
        }
    }
    Ok(())
}
